### ISSUE TITLE
keep-going UX is not good

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#116896]
Anecdotally, I've applied keep-going and it doesn't seem to work. I think this may be because it's really unclear what window of time we actually need to apply keep-going for in order for the CI to pick it up:
- I apply keep-going after the CI test job shows up. Sure, I expect this wouldn't work, but it would be nice if it did.
- I apply keep-going after the CI test job shows up, but resubmits of the PR end up ignoring keep-going. This is what feels like is happening, but I'm not completely sure.

Overall, needing to wait for a round-trip (or more!) for CI to tell me if keep-going is working is a bad UX. We should try to improve this.

cc @pytorch/pytorch-dev-infra

**Comment by huydhn** [comment]
As `keep-going` flag is checked right before the `Test` steps https://github.com/pytorch/pytorch/blob/main/.github/workflows/_linux-test.yml#L131-L140, the flag will be applied if it's there before the test starts.  

Once the test starts, the list of env variables are fixed, so adding `keep-going` at that time won't do anything for the current jobs.  However, if the PR is updated and a new batch of jobs start, they should pick up ``keep-going`

Are these the behavior you observe?

> I apply keep-going after the CI test job shows up. Sure, I expect this wouldn't work, but it would be nice if the test job's keep-going check happened when the first test failed (instead of at the beginning of the job)

Agree that this sounds like a good improvement.

> I apply keep-going after the CI test job shows up, but resubmits of the PR end up ignoring keep-going. This is what feels like is happening, but I'm not completely sure.

This shouldn't be the case and would be a bug otherwise.  If `keep-going` is there before the test step starts, it should be used.  Do you have any example where it behaves differently here?

cc @clee2000

**Comment by zou3519** [comment]
>> I apply keep-going after the CI test job shows up, but resubmits of the PR end up ignoring keep-going. This is what feels like is happening, but I'm not completely sure.

> This shouldn't be the case and would be a bug otherwise. If keep-going is there before the test step starts, it should be used. Do you have any example where it behaves differently here?

Example: https://github.com/pytorch/pytorch/pull/116881

I applied keep-going and submitted an update commit:

![image](https://github.com/pytorch/pytorch/assets/5652049/8e4a3682-6048-469d-af8f-34cf92e47199)

Clicking into the red X, I see the following ([logs](https://github.com/pytorch/pytorch/actions/runs/7425522073/job/20207777942))
![image](https://github.com/pytorch/pytorch/assets/5652049/f5b86e90-5fd4-4651-843f-2bb8fe651493)

**Comment by huydhn** [comment]
I see.  After looking at the log you provide, I think this is a just misleading message from pytest, and `keep-going` flag is indeed used.

So, here is what happened:

1. The message `stopping after 1 failure` came from `pytext -x` flag https://docs.pytest.org/en/7.1.x/how-to/failures.html to stop after the first failure.  We run test with this flag, i.e.
```
'/opt/conda/envs/py_3.11/bin/python', '-bb', 'nn/test_pruning.py', '--shard-id=0', '--num-shards=1', '-v', '-vv', '-rfEX', '-p', 'no:xdist', '--use-pytest', '-x', '--reruns=2', '--import-slow-tests', '--import-disabled-tests'
```

2. The test was rerun 2 more times with `--reruns=2`, and they all failed.  That's why you see a total of 3 `stopping after 1 failure` message for that failed test `test_global_pruning_importance_scores`

3.  With `keep-going` flag, `nn/test_pruning.py` continued as usual after `test_global_pruning_importance_scores` failed https://github.com/pytorch/pytorch/actions/runs/7425522073/job/20207777942#step:16:2665

4. Scrolling down the log, and other tests were also run

cc @clee2000 for another pair of eyes to double check this.

**Comment by clee2000** [comment]
Yes, what huy said re the logs is correct, if you scroll down in the logs you can see that it did end up skipping that test after a bunch of failed reruns, and it continued to run all the tests after it.  The stopping in question just means this instance of pytest stopped.  In general getting results from keep-going relies on doing a lot of staring at logs since it doesn't aggregate the failures that well, especially over multiple test files, and the log classifier on HUD only shows one line right now.  I'll take an AI to 
* improve the log line classification for keep-going
* aggregate failures for keep-going to make finding failures easier

Re the timing for keep-going, I like to put the label on as soon as possible, usually right after I make the PR or while I open the PR

**Comment by zou3519** [comment]
Thanks for checking that keep-going indeed actually works, @huydhn and @clee2000!

> Re the timing for keep-going, I like to put the label on as soon as possible, usually right after I make the PR or while I open the PR

How bad would it be if I wrote a bot to apply keep-going to all of my PRs lol

**Comment by huydhn** [comment]
If there is a group of files that you think having `keep-going` label are needed most of the time, adding a rule for that into https://github.com/pytorch/pytorch/blob/main/.github/labeler.yml is easier.

On the other hand, checking for specific GH usename sounds a bit more complex and could be implement in https://github.com/pytorch/test-infra/blob/main/torchci/lib/bot/autoLabelBot.ts FYI.

**Comment by zou3519** [comment]
Btw, I'm not the only one confused about the timing of keep-going. Could we make it more explicit somehow?

e.g. applying keep-going could restart all the test jobs that have already run.

**Comment by huydhn** [comment]
AI: This would require a design discussion on how to implement it correctly.  Ideally, we want to dynamically apply the behavior when the labels are updated.
